File: c-cpp/med/q22.cpp
/**
 * You are a cognitive scientist, trying to evaluate different schools of thought;
 * through multiple neural pathways called "Synapse"(s).
 * 
 * Each thought has a description, that you give to identify it. (string)
 * Each thought has a significance level, that you give to identify. (double)
 * Each thought has a set of memories associated with it. (list of strings)
 * 
 * You are researching about the history of Philosophy, and you need to evaluate the significance of different
 * schools of thoughts across philosophy, with their associated concepts; that hold the following significance:
 * 
 * 10 Significance: Enhancement Concept
 * 20 Significance: Main Idea
 * 
 * Since you are a cognitive scientist, You've opted to think in terms of neural-networks.
 * 
 * TASK:
 * Your job here, is to create a synaptic pathway for all neurons, such that you could evaluate
 * different schools of thoughts by decorating them with their known significances.
 * 
 * PATTERN:
 * The Design Pattern used here is "Decorator".
 * 
 * Implement the Synapse class to act as the base decorator for extending the functionality of the Thought class. 
 * Ensure that the decorator classes inherit from Synapse and modify both the getDescription() and 
 * getSignificance() methods. 
 * 
 * The code should produce the expected output as described.
 * 
 */


#include <iostream>
#include <vector>
#include <string>
using namespace std;


class Thought {
public:
    string description;  // The description of the thought
    double significance; // The 'significance' or weight of the thought
    vector<string> memory;  // Memories associated

    Thought() {
        description = "Basic Thought";
        significance = 0;  // The initial significance of the thought is neutral
    }

    string getDescription() {
        return description;
    }

    double getSignificance() {
        return significance;
    }

    string getMemory() {
        string memoryStr = "";
        for (const auto& mem : memory) {
            memoryStr += mem + ", ";
        }
        return memoryStr;
    }

    void addMemory(string newMemory) {
        memory.push_back(newMemory);
    }
};


// Synapse Class here


// Main Neuron Decorator, Enhance the thought with MAIN idea.
class MainNeuronDecorator : public Synapse {
public:
    string coreIdea;

    MainNeuronDecorator(Thought* thought, string coreIdea) : Synapse(thought) {
        this->coreIdea = coreIdea;
    }

    string getDescription() {
        return thought->getDescription() + ", Main Idea: " + coreIdea;
    }

    double getSignificance() {
        return thought->getSignificance() + 20;  // Main Idea adds significant weight to the thought
    }
};

// Enhance the thoughts with a CONCEPT!
class EnhancementNeuron : public Synapse {
public:
    string concept;

    EnhancementNeuron(Thought* thought, string concept) : Synapse(thought) {this->concept = concept;}

    string getDescription() {
        return thought->getDescription() + ", Enhanced with " + concept;
    }

    double getSignificance() {
        return thought->getSignificance() + 10;  // Enhances the thought with extra complexity
    }
};



int main() {
    Thought* basicThought = new Thought();  // starting with a neutral thought
    basicThought = new MainNeuronDecorator(basicThought, "Post Structuralism"); // 20
    basicThought = new EnhancementNeuron(basicThought, "Linguistic Ambiguity");  // 10
    basicThought = new EnhancementNeuron(basicThought, "Semantic Brute Permutations"); // 10
    basicThought = new EnhancementNeuron(basicThought, "Concept of Difference"); // 10
    basicThought = new EnhancementNeuron(basicThought, "Concept of Repetition"); // 10
    basicThought = new MainNeuronDecorator(basicThought, "Platonism"); // 20

    basicThought->addMemory("Deleuzean Essay by Daniel W. Smith");  
    basicThought->addMemory("Plato's Republic, Desmond Lee");  


    cout << "Thought Description: " << basicThought->getDescription() << endl;  
    cout << "Thought Significance: " << basicThought->getSignificance() << endl;
    cout << "Thought Memories: " << basicThought->getMemory() << endl;  

    return 0;

    /**
         * EXAMPLE OUTPUT:
            Thought Description: Basic Thought, Main Idea: Post Structuralism, Enhanced with Linguistic Ambiguity, Enhanced with Semantic Brute Permutations, Enhanced with Concept of Difference, Enhanced with Concept of Repetition, Main Idea: Platonism
            Thought Significance: 80
            Thought Memories: Deleuzean Essay by Daniel W. Smith, Plato's Republic, Desmond Lee
    */
}

File: js/med/q22.js
/**
 * You are a cognitive scientist, trying to evaluate different schools of thought;
 * through multiple neural pathways called "Synapse"(s).
 * 
 * Each thought has a description, that you give to identify it. (string)
 * Each thought has a significance level, that you give to identify. (double)
 * Each thought has a set of memories associated with it. (list of strings)
 * 
 * You are researching about the history of Philosophy, and you need to evaluate the significance of different
 * schools of thoughts across philosophy, with their associated concepts; that hold the following significance:
 * 
 * 10 Significance: Enhancement Concept
 * 20 Significance: Main Idea
 * 
 * Since you are a cognitive scientist, You've opted to think in terms of neural-networks.
 * 
 * TASK:
 * Your job here, is to create a synaptic pathway for all neurons, such that you could evaluate
 * different schools of thoughts by decorating them with their known significances.
 * 
 * PATTERN:
 * The Design Pattern used here is "Decorator".
 *
 */

class Thought {
    constructor() {
        this.description = "Basic Thought";
        this.significance = 0;
        this.memory = [];
    }

    getDescription() {
        return this.description;
    }

    getSignificance() {
        return this.significance;
    }

    getMemory() {
        return this.memory.join(", ");
    }

    addMemory(newMemory) {
        this.memory.push(newMemory);
    }
}

// Synapse Class here


// Main Neuron Decorator, Enhance the thought with MAIN idea.
class MainNeuronDecorator extends Synapse {
    constructor(thought, coreIdea) {
        super(thought);
        this.coreIdea = coreIdea;
    }

    getDescription() {
        return this.thought.getDescription() + ", Main Idea: " + this.coreIdea;
    }

    getSignificance() {
        return this.thought.getSignificance() + 20;  // main Idea adds significant weight to the thought
    }
}

// Enhance the thoughts with a CONCEPT!
class EnhancementNeuron extends Synapse {
    constructor(thought, concept) {
        super(thought);
        this.concept = concept;
    }

    getDescription() {
        return this.thought.getDescription() + ", Enhanced with " + this.concept;
    }

    getSignificance() {
        return this.thought.getSignificance() + 10;  // enhances the thought with extra complexity
    }
}

const basicThought = new Thought();  // starting with a neutral thought
let decoratedThought = new MainNeuronDecorator(basicThought, "Post Structuralism"); // 20
decoratedThought = new EnhancementNeuron(decoratedThought, "Linguistic Ambiguity");  // 10
decoratedThought = new EnhancementNeuron(decoratedThought, "Semantic Brute Permutations"); // 10
decoratedThought = new EnhancementNeuron(decoratedThought, "Concept of Difference"); // 10
decoratedThought = new EnhancementNeuron(decoratedThought, "Concept of Repetition"); // 10
decoratedThought = new MainNeuronDecorator(decoratedThought, "Platonism"); // 20

decoratedThought.addMemory("Deleuzean Essay by Daniel W. Smith");  
decoratedThought.addMemory("Plato's Republic, Desmond Lee");  

console.log("Thought Description: " + decoratedThought.getDescription());  
console.log("Thought Significance: " + decoratedThought.getSignificance());
console.log("Thought Memories: " + decoratedThought.getMemory());

/**
         * EXAMPLE OUTPUT:
            Thought Description: Basic Thought, Main Idea: Post Structuralism, Enhanced with Linguistic Ambiguity, Enhanced with Semantic Brute Permutations, Enhanced with Concept of Difference, Enhanced with Concept of Repetition, Main Idea: Platonism
            Thought Significance: 80
            Thought Memories: Deleuzean Essay by Daniel W. Smith, Plato's Republic, Desmond Lee
*/
